### ISSUE TITLE
Split utilities from AbstractRecursivePass out to a trait for easier reuse

### ISSUE THREAD

**Issue by Crell** [symfony/symfony#41726]
**Description**

`Symfony\Component\DependencyInjection\AbstractRecursivePass` includes some very nice utility methods, namely `getReflectionMethod()` and `getConstructor()`.  We want to use those in TYPO3 (which uses Symfony DI), but extending `AbstractRecursivePass` to do something completely non-recursive seems wrong.

Those methods should be simple enough to split out to a trait, though, so that we can use those in another compiler pass without needing inheritance.  I'm happy to file a PR to do so, but I wanted to check first to see if that would be a welcome change.  Though, one complicating factor is the references to $this->currentId, which is tied into the recursion logic, I think?  Suggestions on how to handle that welcome.

There may be other useful utility methods lying about, but these are the ones I'm interested in specifically at the moment.

So, yay/nay on me filing such a PR?

**Comment by stof** [comment]
>  Though, one complicating factor is the references to $this->currentId, which is tied into the recursion logic, I think? Suggestions on how to handle that welcome.

This id is used in a read-only way in those methods, for error reporting. So when extracting an utility, it could receive the service id as argument.

**Comment by ro0NL** [comment]
i tend to prefer `ContainerBuilder::getReflectionMethod` (given we have getReflectionClass already),

for getConstructor i'd be fine with `getReflectionMethod($d, '__construct)` as well

im not sure we should force users to discover a trait first

**Comment by Crell** [comment]
Currently, `getReflectionMethod()` subcalls to `getConstructor()` if the method is a constructor.  I don't really understand why, frankly, but I wasn't going to mess with that logic and leave it essentially intact.  Someone had a reason for doing all that extra work for constructors, but didn't document it obviously.

Moving the logic to `ContainerBuilder` seems better though, I agree, since it's all container-specific logic and `ContainerBuilder` is going to already be available in `process()` anyway.

**Comment by Crell** [comment]
So is that a `yay` on me filing an appropriate PR?

**Comment by nicolas-grekas** [comment]
Should the pass you want really be non-recursive?
Don't you need to process nested inline definition?

**Comment by Crell** [comment]
In our use case, no.  What I ended up doing for now is this: https://git.typo3.org/Packages/TYPO3.CMS.git/blob/HEAD:/typo3/sysext/core/Classes/DependencyInjection/ListenerProviderPass.php

But it seems useful to be able to push essentially that `getRefelctionMethod()` (and maybe `getParameterType()`?) up to ContainerBuilder so that they're more usable by everyone, and we can remove that method.

If not, not biggie, I already solved it for us, but I'd rather solve it for everyone.
